src/preprocess.rkt

```diff
@@ -7,95 +7,79 @@
-      ;; TODO: Handle case where neg isn't supported for this representation
-      (define negate (get-parametric-operator 'neg representation))
-      (replace-vars (list (cons variable (list negate variable))) expression)))
-  (define pairs (combinations variables 2))
-  (define swaps*
-    (for/list ([pair (in-list pairs)])
-      (match-define (list a b) pair)
-      (replace-vars (list (cons a b) (cons b a)) expression)))
-  (define query (make-egg-query (cons expression (append evens* swaps*)) rules))
-  (match-define (cons expression~ rest~) (map last (simplify-batch query)))
-  (define-values (evens~ swaps~) (split-at rest~ (length evens*)))
-  (define swaps (filter-map
-                 (lambda (pair swap~) (and (equal? expression~ swap~) pair))
-                 pairs
-                 swaps~))
-  (define components (connected-components variables swaps))
-  (define abs-instructions
-    (for/list ([variable (in-list variables)] [even~ (in-list evens~)]
-               #:when (equal? expression~ even~))
-      (list 'abs variable)))
-  (define sort-instructions
-    (for/list ([component (in-list components)]
-               #:when (> (length component) 1))
-      (cons 'sort component)))
-  ;; Absolute value should happen before sorting
-  (append abs-instructions sort-instructions))
 
-(define (connected-components variables swaps)
-  (define components (disjoint-set (length variables)))
-  (for ([swap (in-list swaps)])
-    (match-define (list a b) swap)
-    (disjoint-set-union! components
-                         (disjoint-set-find! components (index-of variables a))
-                         (disjoint-set-find! components (index-of variables b))))
-  (group-by
-   (compose (curry disjoint-set-find! components) (curry index-of variables))
-   variables))
+;; TODO: Remove
+;; tests : Expression -> Context -> List (Expression, A)
+;; instructions : List A -> Context -> Instruction
+;; operator : Instruction -> (Point -> Point)
+(struct preprocessing-step (tests instructions operator) #:transparent)
+
+(define (find-preprocessing initial specification context rules node-limit initial-simplify?)
+  (define egraph (make-egraph))
+  (define specification-id (egraph-add-expr egraph specification))
+  (define initial-id (and initial-simplify? (egraph-add-expr egraph initial)))
+  (define test-groups
+    (for/list ([step (in-dict-values preprocessing-steps)])
+      (match-define (preprocessing-step tests _ _) step)
+      (for/list ([test (in-list (tests specification context))])
+        (match-define (cons specification* item) test)
+        (cons (egraph-add-expr egraph specification*) item))))
+  (define (equivalent? id)
+    (= (egraph-find egraph specification-id)
+       (egraph-find egraph id)))
+  (match-define (list _ ffi-rules _) (expand-rules rules))
+  ;; TODO: const-folding?, iter-limit
+  (define iteration-data (egraph-run egraph node-limit ffi-rules #t #f))
```

Comment:
You made this code like 3x more complex so that you could use egraph-find instead of just calling simplify. Bad call.